### PR TITLE
feat(aws-chatbot): Allow adding SNS topic to already existing slack channel

### PR DESCRIPTION
**Pull request by bzbislawski**
Closes [21171](https://github.com/aws/aws-cdk/issues/21171)

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [x] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [x] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by bzbislawski** [comment]
Could I ask for a help with running integration tests and creating new snapshots please?

**Comment by TheRealAmazonKendra** [comment]
@Mergifyio update

**Comment by Naumel** [comment]
Hi, if you need a reference on how to tackle the integ tests, make sure you check out https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md#how-to-write-integration-tests.

**Comment by DanilAgafonov** [comment]
Hello!
@TheRealAmazonKendra @Naumel Is this PR ready to be merged? Could you please take a look?

File: packages/@aws-cdk/aws-chatbot/test/slack-channel-configuration.test.ts [hunk 1]
```diff
@@ -118,6 +118,16 @@ describe('SlackChannelConfiguration', () => {
...
   });
 
+  test('allows adding a Topic to the existing SlackChannel', () => {
+    const slackChannelFromArn = chatbot.SlackChannelConfiguration.fromSlackChannelConfigurationArn(stack, 'RetrievedSlackChannel', 'arn:aws:chatbot::1234567890:chat-configuration/slack-channel/my-slack');
+
+    const topic = new sns.Topic(stack, 'MyTopic');
+
+    slackChannelFromArn.addNotificationTopic(topic);
+
+    Template.fromStack(stack).resourceCountIs('AWS::Chatbot::SlackChannelConfiguration', 0);
```
Comment (marbones-diligent) [inline packages/@aws-cdk/aws-chatbot/test/slack-channel-configuration.test.ts:128]: Is this correct? Shouldn't you check that the SnsTopic is there in the stack? Something like the test above?

Comment (bzbislawski) [reply packages/@aws-cdk/aws-chatbot/test/slack-channel-configuration.test.ts:128]: thanks, updated now

File: packages/@aws-cdk/aws-chatbot/test/integ.chatbot-from-arn.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+      slackChannelId: 'C0187JABUE9', // modify to your slack channel id
+      loggingLevel: chatbot.LoggingLevel.NONE,
+    });
+
+    const slackChannelFromArn = chatbot.SlackChannelConfiguration.fromSlackChannelConfigurationArn(this, 'RetrievedSlackChannel', slackChannel.slackChannelConfigurationArn);
+
+    const topic = new sns.Topic(this, 'MyTopic');
+
+    slackChannelFromArn.addNotificationTopic(topic);
+  }
```
Comment (Naumel) [inline packages/@aws-cdk/aws-chatbot/test/integ.chatbot-from-arn.ts:21]:
```suggestion
    slackChannelFromArn.addNotificationTopic(new sns.Topic(this, 'MyFirstTopic'));
    slackChannelFromArn.addNotificationTopic(new sns.Topic(this, 'MySecondTopic'));
  }
```

File: packages/@aws-cdk/aws-chatbot/test/slack-channel-configuration.test.ts (hunk 1 above)
Comment (Naumel) [reply packages/@aws-cdk/aws-chatbot/test/slack-channel-configuration.test.ts:128]: Perhaps my coffee has not kicked in yet, but how is
`Template.fromStack(stack).resourceCountIs('AWS::Chatbot::SlackChannelConfiguration', 0);`
relevant to the changes/ test?

**Review by Naumel — DISMISSED** [review]
Thank you for your contribution!

Good overall shape, a couple of suggestions with regards to account length in the used ARNs.

**Comment by bzbislawski** [comment]
Maybe something that can help troubleshoot.
Firstly I run `yarn build+test+extract --update-on-failed`
That command asked me to run `yarn integ-runner --update-on-failed`
Here is the error message which I have no clue how to resolve:

`yarn run v1.22.18
$ /Users/zbislawskib/Developer/aws-cdk/node_modules/.bin/integ-runner --update-on-failed

Verifying integration test snapshots...

  NEW        integ.chatbot-from-arn 0.314s
  UNCHANGED  integ.chatbot 1.608s
  UNCHANGED  integ.chatbot-logretention 1.673s

Snapshot Results: 

Tests:    1 failed, 3 total

Running integration tests for failed tests...

Running in parallel across regions: us-east-1, us-east-2, us-west-2
Running test /Users/zbislawskib/Developer/aws-cdk/packages/@aws-cdk/aws-chatbot/test/integ.chatbot-from-arn.js in us-east-1
/Users/zbislawskib/Developer/aws-cdk/packages/@aws-cdk/core/lib/private/refs.js:146
            throw e;
            ^

Error: Resolution error: Resolution error: Resolution error: Resolution error: Context value '@aws-cdk/core:target-partitions' should be a list of strings, got: "@aws-cdk/core:target-partitions".
Object creation stack:
  at stack traces disabled.
Object creation stack:
  at stack traces disabled..
    at ChatbotFromArnInteg.regionalFact (/Users/zbislawskib/Developer/aws-cdk/packages/@aws-cdk/core/lib/stack.js:651:19)
    at ServicePrincipalToken.resolve (/Users/zbislawskib/Developer/aws-cdk/packages/@aws-cdk/aws-iam/lib/principals.js:855:22)
    at RememberingTokenResolver.resolveToken (/Users/zbislawskib/Developer/aws-cdk/packages/@aws-cdk/core/lib/resolvable.js:64:30)
    at RememberingTokenResolver.resolveToken (/Users/zbislawskib/Developer/aws-cdk/packages/@aws-cdk/core/lib/private/resolve.js:198:22)
    at resolve (/Users/zbislawskib/Developer/aws-cdk/packages/@aws-cdk/core/lib/private/resolve.js:131:55)
    at Object.resolve [as mapToken] (/Users/zbislawskib/Developer/aws-cdk/packages/@aws-cdk/core/lib/private/resolve.js:58:48)
    at TokenizedStringFragments.mapTokens (/Users/zbislawskib/Developer/aws-cdk/packages/@aws-cdk/core/lib/string-fragments.js:78:43)
    at RememberingTokenResolver.resolveString (/Users/zbislawskib/Developer/aws-cdk/packages/@aws-cdk/core/lib/resolvable.js:93:26)
    at RememberingTokenResolver.resolveString (/Users/zbislawskib/Developer/aws-cdk/packages/@aws-cdk/core/lib/private/resolve.js:201:27)
    at resolve (/Users/zbislawskib/Developer/aws-cdk/packages/@aws-cdk/core/lib/private/resolve.js:98:54)

Subprocess exited with error 1
  FAILED     integ.chatbot-from-arn-ChatbotFromArn/DefaultTest (undefined/us-east-1) 2.118s
      Integration test failed: Error: Command exited with status 1

Test Results: 

Tests:    1 failed, 1 total
Error: Some integration tests failed!
    at main (/Users/zbislawskib/Developer/aws-cdk/packages/@aws-cdk/integ-runner/lib/cli.js:105:23)
/Users/zbislawskib/Developer/aws-cdk/packages/@aws-cdk/core/lib/private/refs.js:146
            throw e;
            ^

Error: Resolution error: Resolution error: Resolution error: Resolution error: Context value '@aws-cdk/core:target-partitions' should be a list of strings, got: "@aws-cdk/core:target-partitions".
Object creation stack:
  at stack traces disabled.
Object creation stack:
  at stack traces disabled..
    at ChatbotFromArnInteg.regionalFact (/Users/zbislawskib/Developer/aws-cdk/packages/@aws-cdk/core/lib/stack.js:651:19)
    at ServicePrincipalToken.resolve (/Users/zbislawskib/Developer/aws-cdk/packages/@aws-cdk/aws-iam/lib/principals.js:855:22)
    at RememberingTokenResolver.resolveToken (/Users/zbislawskib/Developer/aws-cdk/packages/@aws-cdk/core/lib/resolvable.js:64:30)
    at RememberingTokenResolver.resolveToken (/Users/zbislawskib/Developer/aws-cdk/packages/@aws-cdk/core/lib/private/resolve.js:198:22)
    at resolve (/Users/zbislawskib/Developer/aws-cdk/packages/@aws-cdk/core/lib/private/resolve.js:131:55)
    at Object.resolve [as mapToken] (/Users/zbislawskib/Developer/aws-cdk/packages/@aws-cdk/core/lib/private/resolve.js:58:48)
    at TokenizedStringFragments.mapTokens (/Users/zbislawskib/Developer/aws-cdk/packages/@aws-cdk/core/lib/string-fragments.js:78:43)
    at RememberingTokenResolver.resolveString (/Users/zbislawskib/Developer/aws-cdk/packages/@aws-cdk/core/lib/resolvable.js:93:26)
    at RememberingTokenResolver.resolveString (/Users/zbislawskib/Developer/aws-cdk/packages/@aws-cdk/core/lib/private/resolve.js:201:27)
    at resolve (/Users/zbislawskib/Developer/aws-cdk/packages/@aws-cdk/core/lib/private/resolve.js:98:54)

Subprocess exited with error 1
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.`

**Comment by Naumel** [comment]
@Mergifyio update

File: packages/@aws-cdk/aws-chatbot/test/slack-channel-configuration.test.ts [hunk 3]
```diff
@@ -118,6 +118,16 @@ describe('SlackChannelConfiguration', () => {
...
   });
 
+  test('allows adding a Topic to the existing SlackChannel', () => {
+    const slackChannelFromArn = chatbot.SlackChannelConfiguration.fromSlackChannelConfigurationArn(stack, 'RetrievedSlackChannel', 'arn:aws:chatbot::123456789012:chat-configuration/slack-channel/my-slack');
+
+    const topic = new sns.Topic(stack, 'MyTopic');
+
+    slackChannelFromArn.addNotificationTopic(topic);
+
+    Template.fromStack(stack).resourceCountIs('AWS::SNS::Topic', 1);
```
Comment (mskrip) [inline packages/@aws-cdk/aws-chatbot/test/slack-channel-configuration.test.ts:128]: I may be misunderstanding how this works, but this test case only tests that a `Topic` resource is created. It does not check that there is an association between the topic and the imported chatbot configuration.

File: packages/@aws-cdk/aws-chatbot/test/integ.chatbot-from-arn.js.snapshot/ChatbotFromArnInteg.template.json [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+      "MySlackChannelConfigurationRole1D3F23AE",
+      "Arn"
+     ]
+    },
+    "SlackChannelId": "C0187JABUE9",
+    "SlackWorkspaceId": "T49239U4W",
+    "LoggingLevel": "NONE"
+   }
+  },
+  "MyTopic86869434": {
```
Comment (bjornhandersson) [inline packages/@aws-cdk/aws-chatbot/test/integ.chatbot-from-arn.js.snapshot/ChatbotFromArnInteg.template.json:35]: First, I'm not an expert.
It looks like the snapshots that the test runs against is from this commit [ed8373850e](https://github.com/bzbislawski/aws-cdk/commit/ed8373850e20ee6398d11ba93898a5d6422b2496) where the topic(s) is not the same as for the current state which causes the __integ.chatbot-from-arn__ test to fail.

Can't we just re-generate the snapshot template by remove `test/integ.chatbot-from-arn.js.snapshot/` and re-run the test to generate the new snapshot?
This is the base snapshot so it should be OK right?

I "fixed" this by:
1. Delete `test/integ.chatbot-from-arn.js.snapshot/`
2. `yarn test` -> fail
3. Rename the snapshot folder that was generated from `test/cdk-integ.out.integ.chatbot-from-arn.js.snapshot` to `itest/nteg.chatbot-from-arn.js.snapshot`
4. `yarn test` -> success

File: packages/@aws-cdk/aws-chatbot/test/slack-channel-configuration.test.ts (hunk 3 above)
Comment (sercantor) [reply packages/@aws-cdk/aws-chatbot/test/slack-channel-configuration.test.ts:128]: That is true 👍 I've tested this now and the in the template that's created, there's no reference to [SnsTopicArns](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-chatbot-slackchannelconfiguration.html#cfn-chatbot-slackchannelconfiguration-snstopicarns).

**Pull request closed without merging**